control-plane/agents/src/bin/core/controller/scheduling/volume_policy/thick.rs

```diff
@@ -1,25 +1,102 @@
 use crate::controller::scheduling::{
-    volume::AddVolumeReplica,
-    volume_policy::{
-        pool::{PoolBaseFilters, PoolBaseSorters},
-        DefaultBasePolicy,
-    },
-    ResourceFilter, ResourcePolicy,
+    resources::PoolItem,
+    volume::{AddVolumeReplica, GetSuitablePoolsContext},
+    volume_policy::{pool::PoolBaseFilters, volume_group, DefaultBasePolicy},
+    ResourceFilter, ResourcePolicy, SortBuilder, SortCriteria,
 };
+use std::cmp::Ordering;
+use weighted_scoring::{Criteria, Ranged, ValueGrading};
 
+const TOTAL_REPLICA_COUNT_WEIGHT: Ranged = Ranged::new_const(40);
+const FREE_SPACE_WEIGHT: Ranged = Ranged::new_const(60);
+// Only for volume group multiple replica case.
+const VG_TOTAL_REPLICA_COUNT_WEIGHT: Ranged = Ranged::new_const(25);
+const VG_REPL_COUNT_WEIGHT: Ranged = Ranged::new_const(15);
+
+/// Policy for thick provisioned volumes.
 pub(crate) struct ThickPolicy {}
-impl ThickPolicy {
-    pub(crate) fn new() -> Self {
-        Self {}
-    }
-}
 
 #[async_trait::async_trait(?Send)]
 impl ResourcePolicy<AddVolumeReplica> for ThickPolicy {
     fn apply(self, to: AddVolumeReplica) -> AddVolumeReplica {
         DefaultBasePolicy::filter(to)
             .filter(PoolBaseFilters::min_free_space_full_rebuild)
+            .filter(volume_group::SingleReplicaPolicy::replica_anti_affinity)
             // sort pools in order of preference (from least to most number of replicas)
-            .sort(PoolBaseSorters::sort_by_replica_count)
+            .sort_ctx(ThickPolicy::sort_by_weights)
+    }
+}
+
+impl ThickPolicy {
+    /// Create a new thick policy.
+    pub(crate) fn new() -> Self {
+        Self {}
+    }
+    /// SortCriteria for free space on pool.
+    fn free_space() -> SortCriteria {
```

Comment:
this Criteria are the same as the ones in the thin policy, can we move them to mod.rs for example?

Reply:
No, they cannot be moved yet, because the thick one doesn't have the overcommit yet. So the weights of free space in thick is higher atm, once overcommit is added to thick, we can moved them to common code base. Would take it up later.